File: sqliteReader.py
"""
Read a sqlite3 table.
No error checking.
overflow of payload data not handled

to use:
init the instance with the file name
find_call(<call>)
"""
import struct

def varint(s, ofs):
    r = 0
    i = 0
    j = s[ofs]
    while (j & 128) and (i < 9):
        r = (r << 7) | (j & 127)
        i += 1
        j = s[ofs + i]
    return (r << (8 if i == 8 else 7)) + j, i + 1


class SqliteReader:
    
    #full header decode format ">16sH6B12I20sII"
    #full page header decode format ">B3HB"
    
    def __init__(self, filename):
        self.filename = filename
        recbuffer = bytearray(100)
        with open(filename, "rb") as f:
            f.readinto(recbuffer)
            (self.page_size,
             dummy,
             reserved_space) = struct.unpack_from(">2HB", recbuffer, 16)
            if self.page_size == 1:
                self.page_size = 65536
            self.U = self.page_size - reserved_space # - hdr_ofs?
            self.M = ((self.U - 12) * 32 // 255) - 23
            del recbuffer
            self.recbuffer = bytearray(self.page_size - 100)
            f.readinto(self.recbuffer)
        self._parse_page()
        self.roots = tuple([self._parse_payload(i-100,4) for i in self.cell_ofs])            
        del self.recbuffer
        self.recbuffer = bytearray(self.page_size)
        
    def _read_page(self, f, i):
        f.seek((i - 1) * self.page_size)
        f.readinto(self.recbuffer)
        self._parse_page()
            
    def _parse_page(self):
        offset = 0
        (self.type,
         dummy,
         self.number_cells) = struct.unpack_from(">B2H", self.recbuffer, offset)
        offset += 8
        if self.type in (2, 5):
            self.right_child = struct.unpack_from(">I", self.recbuffer, offset)[0]
            offset += 4
        if self.type != 5:
            self.X = self.U - 35 if self.type == 0x0d else int(((self.U - 12) * 64 // 255) - 23)            
        self.cell_ofs = struct.unpack_from(">%dH" % self.number_cells, self.recbuffer, offset)

    def _parse_payload(self, i, max=None):
        if self.type in (2,5):
            self.left_child = (struct.unpack_from(">I", self.recbuffer, i)[0])
            i += 4
        if self.type != 0x05:
            P, s = varint(self.recbuffer, i)
            i += s
            if P <= self.X:
                base_payload_size = P
            else:
                K = self.M + ((P - self.M) % (self.U - 4))
                base_payload_size = K if K < self.X else self.M
        if self.type in (0x0d, 0x05):
            self.row_id, s = varint(self.recbuffer, i)
            i += s
        if self.type != 0x05:
            overflow = 0 if base_payload_size == P else struct.unpack_from(">I", self.recbuffer, i + base_payload_size)[0]
            self.payload = self._parse_record(i, max)
            return self.payload
        
    def _parse_record(self, ofs, max):
        offset = ofs
        header_size,s = varint(self.recbuffer, offset)
        offset += s
        data_offset = header_size + ofs
        values = []
        ctr = 0
        while (offset < header_size + ofs):
            if ctr == max:
                break
            ctr += 1
            i,s = varint(self.recbuffer, offset)
            offset += s
            if i == 0:  # NULL
                values.append(None)
            elif i == 1:  # 8-bit signed
                values.append(struct.unpack_from("b", self.recbuffer, data_offset)[0])
                data_offset += 1
            elif i == 2:  # 16-bit signed
                values.append(struct.unpack_from(">h", self.recbuffer, data_offset)[0])
                data_offset += 2
            elif i == 3:  # 24-bit signed
                v = struct.unpack_from(">hB", self.recbuffer, data_offset)
                values.append((v[0] << 8) + v[1])
                data_offset += 3
            elif i == 4:  # 32-bit signed
                values.append(struct.unpack_from(">i", self.recbuffer, data_offset)[0])
                data_offset += 4
            elif i == 5:  # 48-bit signed
                v = struct.unpack_from(">iH", self.recbuffer, data_offset)
                values.append((v[0] << 32) + v[1])
                data_offset += 6
            elif i == 6:  # 64-bit signed
                values.append(struct.unpack_from(">q", self.recbuffer, data_offset)[0])
                data_offset += 8
            elif i == 7:  # 64-bit float
                values.append(struct.unpack_from(">d", self.recbuffer, data_offset)[0])
                data_offset += 8
            elif i == 8: # 0
                values.append(0)
            elif i == 9: # 1
                values.append(1)
            elif i in (10,11): # invalid
                pass
            elif i & 1: # text
                ct = (i - 13) // 2
                # values.append(struct.unpack("%ds" % ct, r[data_offset:data_offset + ct])[0].decode())
                values.append(struct.unpack_from("%ds" % ct, self.recbuffer, data_offset)[0].decode())
                data_offset += ct
            else: # blob
                ct = (i - 12) // 2
                values.append(struct.unpack_from("%ds" % ct, self.recbuffer, data_offset)[0])
                offset += ct
        return tuple(values)
    
    
    def find(self, call, table):
        for root in self.roots:
            if root[0] == 'table' and root[1] == table:
                table_no = root[3]
                break
        with open(self.filename, "rb") as f:
            self._read_page(f, table_no)
            while (True):
                lo = 0
                hi = self.number_cells
                while lo < hi:
                    mid = (lo + hi) // 2
                    key = self._parse_payload(self.cell_ofs[mid], 1)[0]
                    if key == call:
                        return self._parse_payload(self.cell_ofs[mid])
                    if key < call:
                        lo = mid + 1
                    else:
                        hi = mid
                if self.type != 2:
                    return ()
                if lo == self.number_cells:
                    child = self.right_child
                else:
                    if (lo != mid):
                        self._parse_payload(self.cell_ofs[lo], 1)
                    child = self.left_child
                self._read_page(f, child)

    def find_call(self, call):
        return self.find(call, 'lookup')


if __name__ == "__main__":
    import sys
    if sys.version.find("Micro") >= 0:
        import os
        import sdcard
        import machine
        cs = machine.Pin(22, machine.Pin.OUT)
        spi = machine.SPI(1, baudrate= 10000000, polarity=0, phase=0, bits=8,
                          firstbit=machine.SPI.MSB, sck=machine.Pin(10),
                          mosi=machine.Pin(11), miso=machine.Pin(12))
        sd = sdcard.SDCard(spi,cs)
        os.mount(sd, '/sd')
        # print (os.listdir('/sd'))
        db_file = "/sd/fcc.sqlite"
    else:
        db_file = "fcc.sqlite"
    r = SqliteReader(db_file)

File: sqliteReader.cpp
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "sqliteReader.h"

struct HEADER {
  char header_string[16];
  uint16_t page_size;
  uint8_t write_version;
  uint8_t read_version;
  uint8_t reserved_space;
  uint8_t max_embedded;
  uint8_t min_embedded;
  uint8_t leaf_payload;
  uint32_t change_counter;
  uint32_t size;
  uint32_t first_freelist;
  uint32_t freelist_pages;
  uint32_t schema_cookie;
  uint32_t schema_format;
  uint32_t default_cache;
  uint32_t largest_root;
  uint32_t text_encoding;
  uint32_t user_version;
  uint32_t inc_vacuum;
  uint32_t application_id;
  char reserved[20];
  uint32_t version_valid;
  uint32_t sqlite_version;
};


#define convert32(value) ((0x000000ff & value) << 24) | ((0x0000ff00 & value) << 8) | \
                       ((0x00ff0000 & value) >> 8) | ((0xff000000 & value) >> 24)

#define convert16(value) ((0x00ff & value) << 8) | ((0xff00 & value) >> 8)




sqliteReader::sqliteReader(const char *name) {
  // printf("sizeof HEADER %lu\r\n", sizeof(HEADER));
  filename = (char *) malloc(strlen(name) + 1);
  strcpy(filename, name);
  cell_offset = NULL;

  FILE *f = fopen(filename, "rb");
  HEADER hdr;
  fread(&hdr, sizeof(HEADER), 1, f);
  
  page_size = convert16(hdr.page_size);

  U = page_size - hdr.reserved_space;
  M = ((U - 12) * 32 / 255) - 23;
  
  page_buffer = (uint8_t *) malloc(page_size);

  fread(page_buffer, page_size - sizeof(HEADER), 1, f);
  fclose(f);
  parse_page();
  printf("cell_offset %d\r\n", cell_offset[0]);
  parse_payload(cell_offset[0] - 100);

}

void sqliteReader::parse_page() {
  uint16_t offset = 0;
  type = page_buffer[offset];
  max_number_cells = 0;
  number_cells = page_buffer[offset + 3] << 8 | page_buffer[offset + 4]; 
  offset = 8;
  if (type == 2 || type == 5) {
    right_child = page_buffer[offset] << 24
      | page_buffer[offset + 1] << 16
      | page_buffer[offset + 2] << 8
      | page_buffer[offset + 3];  
      offset += 4;
  }
  if (type != 5) {
    X = (type == 5) ? U - 35 : (U - 12) * 64 / 255 - 23;
  }
  if (number_cells > max_number_cells) {
    free(cell_offset);
    cell_offset = (uint16_t *)malloc(number_cells * 2);
    max_number_cells = number_cells;
  }
  memcpy(cell_offset, page_buffer + offset, number_cells * 2);
  for (int i=0; i<number_cells; i++) {
    cell_offset[i] = convert16(cell_offset[i]);
  }
  
}

void sqliteReader::parse_payload(uint16_t ofs, uint16_t max) {
  uint32_t base_payload_size;
  uint32_t P;
  uint16_t offset = ofs;
  if (type == 2 || type == 5) {
    left_child = page_buffer[offset] << 24
      | page_buffer[offset+1] << 16
      | page_buffer[offset+2] << 8
      | page_buffer[offset+3];  
      offset += 4;
  }
  if (type != 5) {
    P = varint(offset);
    if (P <= X) {
      base_payload_size = P;
    } else {
      uint16_t K = M + ((P - M) % (U - 4));
      base_payload_size = K < X ? K : M;
    }
    // printf("payload size = %u\r\n", P);
    // printf("base payload size = %u\r\n", base_payload_size);
  }
  if (type == 0xd || type == 5) {
    row_id = varint(offset);  
  }
  printf("row_id = %d, offset = %d\r\n", row_id, offset);
  if (type != 5) {
    if (P == base_payload_size ) {
      overflow = 0;
    } else {
      uint16_t of_offset = offset + base_payload_size;
      overflow = page_buffer[of_offset] << 24
        | page_buffer[of_offset+1] << 16
        | page_buffer[of_offset+2] << 8
        | page_buffer[of_offset+3];  
    }
    parse_record(offset);
  }
}

void sqliteReader::parse_record(uint16_t ofs, uint16_t max) {
  uint16_t offset = ofs;
  uint32_t header_size = varint(offset);
  printf("header_size = %u\r\n", header_size);
  uint32_t data_offset = header_size + ofs;
  uint32_t ctr = 0;
  uint32_t field_type;
  int32_t value;
  
  while (offset < header_size + ofs) {
    if (ctr == max) {
      break;
    } 
    field_type = varint(offset);
    printf("field type: %u ", field_type);
    switch (field_type) {
      case 0:
        // printf("NULL\r\n");
        break;
      case 1:
        value = (int8_t)page_buffer[data_offset++];
        // printf("byte %d\r\n", value);
        break;
      case 2:
        value = (int8_t)page_buffer[data_offset] << 8
                      | (int8_t)page_buffer[data_offset + 1];
        data_offset += 2;
        printf ("short %d\r\n", value);
        break;
      default:
        //printf("\r\n");
        break;
    }
  }

}

uint32_t sqliteReader::varint(uint16_t &offset) {
  uint32_t r = 0;
  uint16_t i = 0;
  uint8_t j = page_buffer[offset++];
  while ((j & 0x80) && (i < 9)) {
    r = (r << 7 | j & 0x7f);
    i++;
    j = page_buffer[offset++];
  }
  return r << (i == 8 ? 8 : 7) | j;
}

sqliteReader::~sqliteReader() {
  free(page_buffer);
  free(filename);
  free(cell_offset);
}

int main(void) {
  sqliteReader reader = sqliteReader("fcc.sqlite");
}